012 - Hot reload feature proposal (#83)
kroxylicious/design: proposals/083-hot-reload-feature.md

A configuration apply triggered by an HTTP endpoint should behave identically to one triggered by a file watcher or an operator callback.
These decisions belong in the proxy's static configuration, keeping the `applyConfiguration()` signature simple.

### Configuration change detection

When `applyConfiguration()` is called, the proxy compares the new configuration against the running state to determine which virtual clusters need to be restarted.
A cluster requires a restart if any of the following changed:

- **The virtual cluster model itself** — bootstrap address, TLS settings, gateway configuration, or any other property that contributes to `VirtualClusterModel.equals()`.
- **A filter definition used by the cluster** — if the type or configuration of a `NamedFilterDefinition` referenced by the cluster changed (compared via `equals()`).
- **The default filters list** — if the cluster relies on default filters and the default filters list changed (order matters, since filter chain execution is sequential).

> Clusters where none of these changed are left untouched — they continue serving traffic throughout the apply operation.

Reviewer: The restructured proposal addresses this in the "Plugin resource tracking (known gap)" section — it explicitly calls out that in-place changes to external files won't be detected, and defers the solution to a separate proposal. The compatibility angle of going from ignoring to honouring those changes can travel with that future proposal.

| `onFailure` | `ROLLBACK` | **Default.** Undo all changes, restore proxy to previous stable state.                                 |
| | `TERMINATE` | Shut down the entire proxy. Let external supervision (K8s, systemd) restart.                           |
| `persistConfigToDisk` | `true` | **Default.** Write the new configuration to the config file (with `.bak` backup of the replaced file). |
| | `false` | Don't persist. The reload is in-memory only.                                                           |

### Configuration change detection

When `applyConfiguration()` is called, the proxy compares the new configuration against the running state to determine which virtual clusters need to be restarted. Change detection is implemented as a pipeline of `ChangeDetector` implementations, each responsible for one category of change:

- **`VirtualClusterChangeDetector`** — identifies clusters that were added, removed, or modified by comparing `VirtualClusterModel` instances via `equals()`. A cluster requires a restart if any property that contributes to `VirtualClusterModel.equals()` changed (bootstrap address, TLS settings, gateway configuration, etc.).

> - **`FilterChangeDetector`** — identifies clusters affected by filter configuration changes. A cluster requires a restart if a `NamedFilterDefinition` it references changed (type or configuration, compared via `equals()`), or if the `defaultFilters` list changed (order matters, since filter chain execution is sequential) and the cluster relies on default filters.

Reviewer 2: Essentially, filter config changes will not support hot reload?

Author: Any config changes to filter config will trigger a hot reload.
There are caveats that if there is a change in a content of file (which is referred in filter or virtual cluster config), these detector's wont be able to detect file content changes. This is explicitly called out in this proposal and this is to be solved later.
For eg:- certificate gets rotation but the file name remains same

**Remove:** `SERVING → DRAINING → [drain connections] → [deregister gateways] → STOPPED`

A removed cluster is permanently torn down. It reaches the terminal Stopped state, and the `onVirtualClusterStopped` callback fires. If the proxy's `onVirtualClusterStopped` failure policy is `serve: none` (the default from Proposal 016), the proxy shuts down. If the policy is `serve: successful`, the proxy continues with the remaining healthy virtual clusters.

**Add:** `[create lifecycle manager in INITIALIZING] → [register gateways] → SERVING`

A new cluster starts in the Initializing state with a fresh `VirtualClusterLifecycleManager`, registers its gateways with the `EndpointRegistry`, and transitions to Serving.

Changes are processed in the order: **remove → modify → add**. Removing clusters first frees up ports and resources that new or modified clusters may need.

> This means a modified cluster experiences a brief period of unavailability while its ports are unbound and rebound. Clients connected to the cluster will be disconnected during the drain phase. This is a deliberate design choice. More surgical approaches — such as swapping the filter chain on existing connections without dropping them, or performing a rolling handoff — would reduce disruption, but they add significant complexity. The remove+add approach is the right starting point: it is straightforward, predictable, and consistent with how the proxy handles startup failures today. The remove+add approach also creates a thundering herd when all disconnected clients reconnect simultaneously after the cluster comes back up; mitigation strategies (e.g. staggered connection acceptance) are future work.

Reviewer 2: the unavailability problem isn't tied to removed clusters existing — it comes from how modify itself works (unbind ports →  rebind). Even with zero removals, each modified cluster still goes through SERVING → DRAINING → deregister → INITIALIZING → register → SERVING, causing downtime on that cluster.                                                                            
                 
And if modified clusters are processed sequentially, the total unavailability window compounds — cluster B doesn't start its restart until cluster A's restart completes. The justification for the remove-first ordering is valid (port reuse), but is not the only reason for unavailability.

When a VC traverses the `failed → stopped` edge — meaning it is truly unrecoverable — what's the blast radius? `serve: none` (proxy shuts down) or `serve: successful` (remaining VCs continue). This fires only on the `failed → stopped` edge, not on `draining → stopped` (intentional removal) or `initializing → stopped` (shutdown during startup). The lifecycle state model is unchanged — same states, same transitions. The intelligence is in the `VirtualClusterManager` knowing which transitions are policy-triggering, not in the state itself.

### Configuration model

Failure behaviour is deployment-level static configuration, split into two independent policy dimensions. A configuration apply triggered by an HTTP endpoint should behave identically to one triggered by a file watcher or an operator callback. These decisions belong in the proxy's static configuration, keeping the `applyConfiguration()` signature simple.

```yaml
proxy:
  # Lifecycle: fires on the failed → stopped edge only
  # Applies at startup AND reload
  onVirtualClusterTerminalFailure:
    serve: none              # none | successful

  # Reload-specific settings

>   configurationReload:

Reviewer 3: In terms of the proxy _as it's packaged by the project_, this does not make sense. 

* The project provides the proxy application (and container images containing that). But nothing in the proxy application would actually make use of this configuration because you've defined the scope of the proposal to be the Java method `applyConfiguration()`. You've kept how that method gets called out of scope for the proxy application. Therefore, this is proxy configuration that's not honoured by the proxy.

* My understanding is that you're embedding the proxy. That's not an API that we consider "supported" (in the sense of caring about backwards compatibility between releases). The application you're embedding it into gets to call `applyConfiguration()`, and could use this block to configure how that behaves.

So it seems wrong to me to propose adding a `configurationReload` configuration property when the proxy _as provided by the project_ does not support reload.

Author: "My understanding is that you're embedding the proxy" - yes thats right.
In my earlier design I had proposed a `ReloadOptions` parameter as part of this API, which would be passed by the invoker of this API. But there were some concerns around it - https://github.com/kroxylicious/design/pull/83#issuecomment-3918271213

This was the earlier plan
- Keep `onVirtualClusterTerminalFailure` in YAML. It explicitly applies at startup as well as reload, so the shipped binary already honours it independent of any trigger.
- Drop the `configurationReload`: YAML block entirely. Instead have a `ReloadOptions` parameter as part of the `applyConfiguration` API call, which has the following two properties - `rollback` ,   `persistToDisk`                                                                                             

Any other thoughts on this ? @tombentley @SamBarker

Author: Hi fellas, bumping this up, what do we think would the be right way ahead

Reviewer: `configurationReload.rollback` belongs in the proxy config rather than as a per-call `ReloadOptions` parameter because it is trigger-agnostic: an HTTP trigger and a file watcher trigger should see identical behaviour for the same `newConfig`. The call site shouldn't be able to vary the policy, but the policy can legitimately vary as part of an atomic configuration change.

@tombentley you're right that without a reload trigger wired up, this config is moot. But that's the same position we took with `onVirtualClusterFailed` retries in Proposal 016 — establishing the config entry point now keeps the contract clear and gives future trigger implementations an unambiguous place to plug in.

I'd suggest renaming the block and its field to better reflect what they govern:

```yaml
proxy:
  onVirtualClusterTerminalFailure:
    serve: none | successful
  updateStrategy:          # was: configurationReload
    onFailure: revert | continue   # was: rollback: true | false
```

This keeps `newConfig` self-describing: it contains both the desired end state and the policy for how to get there. The `updateStrategy` in `newConfig` governs the application of that specific configuration — it's part of what that configuration means.

`updateStrategy.onFailure` composes cleanly with `onVirtualClusterTerminalFailure.serve`:

| `onFailure` | `serve`      | Behaviour |
|-------------|--------------|-----------|
| `revert`    | `none`       | Atomic reload. Revert on failure. If revert itself fails, proxy shuts down. |
| `revert`    | `successful` | Atomic reload. Revert on failure. If revert itself fails, surviving VCs continue. |
| `continue`  | `none`       | Press forward, but any dead VC shuts the proxy down. |
| `continue`  | `successful` | Press forward, dead VCs dropped, rest continue. (e.g. TLS enforcement.) |

`persistToDisk` genuinely is a trigger concern — a Kubernetes operator owns config state via CRD and doesn't want the proxy overwriting its file. That belongs to the trigger, not the proxy.

Proposal updates needed:
- Adopt the naming above
- Remove `persistToDisk`
- Add `ReloadOptions` as a per-call parameter to the rejected alternatives, with the reasoning that call sites shouldn't vary the policy

Reviewer 3: @SamBarker 

>  The call site shouldn't be able to vary the policy, but the policy can legitimately vary as part of an atomic configuration change.

I can agree that _would_ be a legitimate thing to do if the proxy itself supported reloading. But the fact is that it does not, so it's wrong to give it a configuration which doesn't do anything thing. And while it's conceptually meaningful to change the policy mid-flight, I think it's practical utility is doubtful.

So I'm much more in favour of a `ReloadOptions`. That keeps the configuration (our API) clean, and doesn't impact the practical usefulness.

Reviewer: For now, let's sidestep this by hardcoding `revert`. It's the safe default —
on failure the proxy always attempts to restore all VCs to the previous
known-good configuration. No config entry, no per-call parameter. The
behaviour is documented in the proposal as a deliberate choice, and we
revisit configurability when a concrete use case demands a second option.

That unblocks the proposal without requiring us to agree on where the knob
lives.

One thing I'm not conceding: when that moment comes, the right home is config
rather than a per-call parameter. `ReloadOptions` allows trigger
implementations to vary the failure policy independently of each other, which
breaks behavioural consistency across triggers.

For now: harden `revert`, remove `updateStrategy` from the proposal, and
move on.

Author: Thanks @tombentley @SamBarker for looking into this.

Unfortunately, hardening the revert wont work for us as our internal plan depends upon the proxy using the 
`onFailure`: `continue` and `serve`: `none` configs, aka we want the proxy to shutdown in case of any VC failures.
We are managing the config reverts on the k8s operator side

> # Changing Active Proxy Configuration

Reviewer 4: ```suggestion
# 83 - Changing Active Proxy Configuration
```

Author: done

In Kubernetes deployments, a configuration change means a pod restart. In standalone deployments, it means stopping and restarting the process. Both cause a service interruption that is disproportionate to the scope of the change.

## Motivation

Administrators need to be able to modify proxy configuration in place. Common scenarios include:

- **Adding or removing virtual clusters** as tenants are onboarded or offboarded.
- **Updating filter configuration** (e.g. updating a KMS endpoint, changing a key selection pattern, modifying ACL rules).

The proxy should apply these changes with minimal disruption: only the virtual clusters affected by the change should experience downtime. Unaffected clusters should continue serving traffic without interruption.

This proposal also delivers the runtime capability that Proposal 016 made possible: with `applyConfiguration()` available, an operator (via whatever trigger they're using) can let healthy VCs continue serving while fixing a broken VC's config and re-applying — making per-VC independence useful rather than theoretical.

## Proposal

### Core API: `KafkaProxy.applyConfiguration()`

The central operation is:

```java

> class KafkaProxy {

Reviewer 4: The KafkaProxy ctor takes a Configuration object already.  Am I right in saying that the initial configuration continues to be supplied from the ctor, and it is only replacement configuration that are supplied by the applyConfiguration method, right?

Author: Yes, exactly, Initial config stays on the ctor.
`applyConfiguration()` handles replacements on a running proxy. I have made it explicit in the Javadoc's Scope subsection.

- Other VCs that were already modified earlier in the apply are not reverted — their new configuration remains in effect.
- Pending changes for other VCs continue to be processed.

The caller receives the full set of failures and chooses the response (shutdown, rollback, alert, retry) via `whenComplete`. Three illustrative trigger-side patterns are shown in the [Core API](#core-api-kafkaproxyapplyconfiguration) section above.

Rollback, in particular, is a trigger-side concern: a trigger that wants atomic "all-or-nothing" semantics calls `applyConfiguration(oldConfig)` from its `whenComplete` callback when the previous apply returned non-empty `errors()`. The proxy does not know it is "rolling back" — it just runs another apply. This keeps the proxy implementation small and lets different deployment scenarios (CLI, operator, custom trigger) express different policies without proxy changes.

### Orchestration pipeline

The complete `KafkaProxy.applyConfiguration()` pipeline flows through these layers:

```
KafkaProxy.applyConfiguration(newConfig)
    │
    ├── Guards: proxy must be running, orchestrator must be initialized
    │
    ▼

> ConfigurationReloadOrchestrator.reload(newConfig)

Reviewer 4: can reload requests queue up? what happens if a reload requests arrives when another is still in flight?

Author: Added a Concurrency control section

Concurrent calls are not queued, second call completes exceptionally with ConcurrentApplyException. Trigger is expected to retry. 
Section also covers the rationale

5. Caller's whenComplete calls applyConfiguration(oldConfig):
     - Removes VC-A's new config, restores old config: VC-A drains and re-initializes
     - Re-adds VC-B with old config: succeeds, VC-B back to serving
6. Rollback's future completes with errors() empty. The system is back to oldConfig.
```

If the rollback itself returns non-empty `errors()`, the trigger applies its last-resort policy (typically shutdown).

#### Catastrophic failure

```
1. applyConfiguration(newConfig) is called.
2. The orchestrator encounters an internal error (e.g. an unexpected runtime exception
   in change-detection itself, not in any particular VC).
3. Future completes exceptionally with that throwable.
4. Caller's whenComplete sees ex != null and applies its catastrophic-error policy.
```

The distinction between "per-VC failure" (future completes normally, `errors()` non-empty) and "catastrophic failure" (future completes exceptionally) is the boundary between *the apply ran and at least one component failed* and *the apply could not be evaluated at all*.

### Metrics and observability

The following metrics are part of the reload implementation:

> - **`kroxylicious_apply_total`** — counter of `KafkaProxy.applyConfiguration()` invocations, labelled by outcome (`success` = future completes with empty `errors()`; `partial_failure` = future completes with non-empty `errors()`; `catastrophic` = future completes exceptionally). Enables alerting on apply failures and tracking apply frequency.

Reviewer 4: I would make these metric names config specific eg. `kroxylicious_apply_config`

Author: done, renamed `kroxylicious_apply_*` → `kroxylicious_apply_config_*`

> - **`kroxylicious_apply_total`** — counter of `KafkaProxy.applyConfiguration()` invocations, labelled by outcome (`success` = future completes with empty `errors()`; `partial_failure` = future completes with non-empty `errors()`; `catastrophic` = future completes exceptionally). Enables alerting on apply failures and tracking apply frequency.

Reviewer 4: I assume that this won't count the initial configuration, so after the first applyConfiguration this counter will be 1, not 2.

Author: Correct, counter only increments on explicit applyConfiguration() calls, the initial startup config does not increment it. Adding that explicitly to the metric description

The proxy should apply these changes with minimal disruption: only the virtual clusters affected by the change should experience downtime. Unaffected clusters should continue serving traffic without interruption.

This proposal also delivers the runtime capability that Proposal 016 made possible: with `applyConfiguration()` available, an operator (via whatever trigger they're using) can let healthy VCs continue serving while fixing a broken VC's config and re-applying — making per-VC independence useful rather than theoretical.

## Proposal

### Core API: `KafkaProxy.applyConfiguration()`

The central operation is:

```java
class KafkaProxy {
  // ... add the following method

  /**
   * Apply the given configuration to this running proxy, restarting only the
   * virtual clusters whose effective configuration differs from the current
   * running state. Unaffected clusters continue serving traffic throughout
   * the apply.
   *
   * <h2>Scope</h2>
   * <p>This method handles <em>replacement</em> configurations on an already-running
   * proxy. The initial configuration continues to be supplied via the {@link KafkaProxy}

>    * constructor at proxy startup; that path is unchanged by this proposal.

Reviewer 3: Remember this is the Javadoc of the code you're proposing to add, so that Javadoc shouldn't talk about "this proposal".

```suggestion
   * proxy. The initial configuration is supplied via the {@link KafkaProxy}
   * constructor at proxy startup.
```

   *
   * <p>Within that scope, this method applies only the virtual-cluster sections of
   * the configuration and the named filter definitions that those virtual clusters
   * reference. Other configuration sections (management, metrics, admin, etc.) are
   * out of scope and are not reconciled by this operation; changes to those sections

>    * still require a proxy restart.

Reviewer 3: When you say "are not reconciled by this operation" does that mean that differences between the existing configuration and the given `newConfig` in those other configuration sections are ignored? Or result in an exceptionally complete future?

Author: I was thinking to ignore them silently, but now that I think of it, if we return an exceptionally complete future, we can immediately do a fail fast and let the operator know.
WDYT ?

Reviewer 3: I think the scope of what is reloadable is already a bit blurry (e.g. that key stores, tls material evades detection). If we silently ignore differences in other section too then that's just blurriness of another kind. 

The problem with the ignoring behaviour is it can be difficult to evolve while maintaining compatibility. A better future implementation (which supported hot reload of some other sections) would suddenly start doing something it didn't before. So there would be a change in semantic. So you'd end up needing some flag to opt into the new behaviour, and it all just ends up being hard for people to understand, with a bigger configuration and testing surface.

For this reason I favour fail fast.

Author: Sounds good, I'll update the doc and will plan to fail and return a exceptionally future with `OutOfScopeChangeException`

   *       The future completes exceptionally.</li>
   *   <li><b>Per-component failure</b> — the apply was evaluated and one or
   *       more components (virtual clusters or referenced filters) failed to
   *       converge. The future completes normally with a {@code ConfigurationResult}
   *       whose {@code errors()} collection is non-empty.</li>
   * </ul>
   *
   * <p>Failure-handling policy (whether to shut down on partial failure, attempt
   * a rollback, alert, or retry) is the caller's responsibility, expressed via
   * the standard {@link java.util.concurrent.CompletableFuture#whenComplete}
   * pattern. The proxy itself takes no policy action based on {@code errors()};
   * it only reports.
   *
   * @param newConfig the desired end-state configuration; must be non-null
   *                  and statically valid
   * @return a future that completes with a {@link ConfigurationResult} describing
   *         any per-component failures encountered while applying the
   *         configuration
   * @throws NullPointerException  if {@code newConfig} is {@code null}
   * @throws IllegalStateException if the proxy is not in the running state
   */

>   public CompletableFuture<ConfigurationResult> applyConfiguration(Configuration newConfig);

Reviewer 3: I know it's a bit late in the day to make this suggestion, but I think "apply configuration" is quite cumbersome terminology.

```suggestion
  public CompletableFuture<ReconfigureResult> reconfigure(Configuration newConfig);
```

Obviously this change would also need to flow through the rest of the propsal, for example it would need to be `ReconfigureError`, and the metric names would change.

Author: I can make this change if you and @SamBarker think its the right terminology, will have to change each and every reference

Reviewer 4: +1 to reconfigure.

Author: This change was a handful, have renamed it to `reconfigure()` and also renamed the associated classes

#### Caller-side failure handling

Because the proxy does not act on `errors()`, callers express their failure policy via `whenComplete` on the returned future. Three illustrative patterns:

**Shut down on any failure**:

```java
proxy.applyConfiguration(newConfig)
     .whenComplete((result, ex) -> {
        if (ex != null) {
        LOGGER.atError().setCause(ex).log("Configuration apply failed catastrophically");
             proxy.shutdown();
             return;
                     }
                     for (var error : result.errors()) {
        LOGGER.atError()
                   .setCause(error.cause())
        .addKeyValue("component", error.humanReadableIdentifier())
        .log("Configuration apply failed for component");
         }
                 if (result.hasErrors()) {
        proxy.shutdown();
         }

>                  });

Reviewer 3: Please can we fix the indentation here.

```

The future completes with the aggregate result *before* any action is taken, so logging happens before shutdown — no ordering problem.

**Best-effort apply, keep what worked**:

```java
proxy.applyConfiguration(newConfig)
     .whenComplete((result, ex) -> {
        if (ex != null) {
        alerter.send("catastrophic-apply-failure", ex);
             return;
                     }
                     for (var error : result.errors()) {
        alerter.send("component-apply-failure", error);
         }
                 // Surviving components continue serving; no proxy-level action taken.
                 });
```

**Rollback on failure** (a sophisticated trigger):

```java
proxy.applyConfiguration(newConfig)
     .whenComplete((result, ex) -> {
        if (result != null && result.hasErrors()) {

>         proxy.applyConfiguration(oldConfig)

Reviewer 3: I assume the expectation is that it's the caller's job to keep track of what the `oldConfig` was? I think that's fine, at least for now. But it would be best to say that somewhere in this proposal.

Author: added

- **`VirtualClusterChangeDetector`** — identifies clusters that were added, removed, or modified by comparing `VirtualClusterModel` instances via `equals()`. A cluster requires a restart if any property that contributes to `VirtualClusterModel.equals()` changed (bootstrap address, TLS settings, gateway configuration, etc.).
- **`FilterChangeDetector`** — identifies clusters affected by filter configuration changes. A cluster requires a restart if a `NamedFilterDefinition` it references changed (type or configuration, compared via `equals()`), or if the `defaultFilters` list changed (order matters, since filter chain execution is sequential) and the cluster relies on default filters.

Detectors return a `ChangeResult(clustersToRemove, clustersToAdd, clustersToModify)`. Results from all detectors are aggregated and then passed onto `VirtualClusterManager` to perform relevant operations.

Clusters where none of these changed are left untouched — they continue serving traffic throughout the apply operation.

### Cluster modification via lifecycle transitions

> A modified virtual cluster is restarted by driving it through the lifecycle states defined in Proposal 016. Proposal 016 defines the per-VC state machine (`VirtualClusterLifecycleState`) and the `VirtualClusterRegistry` that enforces valid transitions. This proposal adds the reload operations that drive those transitions.

Reviewer 3: You're talking about the `VirtualClusterRegistry` and operations: Are those operations new _methods_ on `VirtualClusterRegistry`? 

Can we have a dedicated `replace()` operation, even if it's just a `remove()` and `add()` internally? It would be better to do that from the start than to try to retrospectively add a replace semantic if callers had already been forced to replace as `remove()` and `add()` for themselves. Maybe we'll never get around to a better implementation of replace, but allowing for this now seems like a small cost.

Author: Yes, there will new methods, have update the doc to include all 3 methods - 
`removeVirtualCluster`, `replaceVirtualCluster` and `addVirtualCluster`

Before tearing down a modified or removed cluster, the proxy drives its lifecycle from **Serving to Draining**. The detailed mechanics of connection draining — rejecting new connections, applying backpressure, waiting for in-flight requests, and force-closing after timeout — are defined as part of the `Draining` lifecycle state in Proposal 016 and its implementation. This proposal does not redefine that behaviour; it relies on the lifecycle state machine to handle drain semantics.

Once all connections are drained (or the drain timeout expires), the lifecycle transitions out of Draining:
- For **restart**, gateways are deregistered and re-registered, the lifecycle transitions through Initializing to Serving.
- For **remove**, the lifecycle manager transitions from Draining to Stopped via `drainComplete()`. This is the `draining → stopped` edge — the terminal failure callback does **not** fire.

### FilterChainFactory hot-swap

Filter configuration changes require replacing the `FilterChainFactory` that creates filter chains for new connections. The existing architecture creates `FilterChainFactory` once at startup and passes it as a `final` reference to `KafkaProxyInitializer`.

> To support hot reload, `KafkaProxy` holds an `AtomicReference<FilterChainFactory>` shared between:

Reviewer 3: I'm keen that we don't pass around an `AtomicReference` -- it has a much wider interface than is really required. `KafkaProxyInitializer` just needs a `Supplier`.

Author: I dont understand how a `Supplier` would work here, as we need to "swap" the filter factory and not just get it.

If you mean  to suggest "Supplier-replaced-on-swap" (volatile Supplier field), it might make sense to just use a AtomicReference<> instead.

Reviewer 3: As I understood what you were proposing the KafkaProxyInitializer just needs to get the current `FCF`. So from its PoV a `Supplier` (`atomicReference::get`) would be sufficient. At the point you perform the swap you need to know that it's all underpinned by an `AtomicReference`.

Author: >KafkaProxyInitializer just needs to get the current FCF

That's right, but we need to get the current `FCF`, which was "swapped" by the reload operations. 
Maybe we can keep a `Supplier<FCF>` in KafkaProxyInitializer, but internally we back it up via a `AtomicReference`, but that will just increase the code to maintain the POV semantics.
I don't think we should deep dive into the internals in the design doc, we can discuss on this in the actual PR. Does this work for you ?

Reviewer 3: Sure.

- `KafkaProxyInitializer` — reads via `.get()` on each new connection, always getting the current factory

> - `ConfigurationReloadOrchestrator` — swaps via `.set()` after successful reload

Reviewer 3: This class is mentioned here and again below, but is never really described. 

I don't really want to force you into describing a whole lot of internal implementation detail, but it would be good to add at least a _little_ more detail. For example:
* is it entirely an internal implementation detail of `KafkaProxy`? 
* What is its relationship with the change detectors and the virtual cluster registry?
* For the change detectors, are you proposing that it call those directly itself, or merely acts on a `ChangeResult` is has received from the `KafkaProxy`?

Author: Fair points.

1.  It's a private class of kroxylicious-runtime. KafkaProxy constructs and holds it, embedders never obtain a reference. The sole public entry point is KafkaProxy.applyConfiguration()
2. The orchestrator calls the detectors to gather the `ChangeResult` which holds the list of virtual clusters which are to be either added/removed/restarted. Orchestrator then calls the `VirtualClusterRegistry` to perform the relevant operations.  Endpoint registration and per-VC lifecycle state are not the             orchestrator's concern as they are part of the registry
3. Above point should answer this question.

Having said this, I will go ahead and add a small section about the orchestrator and also update the `Orchestration pipeline` diagram to reflect this.

Rejecting (rather than queuing) the second call is deliberate. Queuing would raise three questions the proposal does not want to answer:

- **Staleness**: a queued apply may carry configuration that is already obsolete by the time it executes (the trigger's source of truth has moved on). Triggers cannot easily detect this.
- **Bounded depth**: an unbounded queue is a memory hazard; a bounded queue forces a "what to do when full?" policy, which is the very thing we're trying to avoid pushing into the proxy.
- **Coalescing semantics**: should an arriving call replace the oldest queued call, the most recent one, or neither? Each choice is defensible and trigger-specific.

By rejecting fast, the proxy forces these decisions onto the trigger, where they can be made with full knowledge of the trigger's source of truth. A trigger that wants "always apply the most recent config" can implement it cleanly via the rejection-and-retry loop (debounce, then call `applyConfiguration` with the latest config); a trigger that needs different semantics can implement those too without the proxy having pre-committed to a policy.

> `ConcurrentApplyException` is the exceptional-completion cause for this scenario (i.e. accessed via `future.exceptionally(...)` or the `ex` parameter of `whenComplete`, not thrown synchronously from `applyConfiguration` itself — that distinction matches the rest of the error-reporting contract).

Reviewer 4: I think the new exception brings a  hazard for the client's error handling.  You said above that a client may follow an algorithm like this to handle rollback.

```
proxy.applyConfiguration(newConfig)
     .whenComplete((result, ex) -> {
         if (result != null && result.hasErrors()) {
             proxy.applyConfiguration(oldConfig)
                  .whenComplete((rollbackResult, rollbackEx) -> {
                      if (rollbackEx != null || rollbackResult.hasErrors()) {
                          // rollback itself failed — last-resort policy
                          proxy.shutdown();
                      }
                  });
         }
     });
```

Don't we need to special case ConcurrentApplyException?  If no, a pair of client threads that are racing to update the proxy config, the one that receives a ConcurrentApplyException may inadvertently rollback the other client's applied changes.

Author: Great catch.

This can be easily handled by checking for `ConcurrentApplyException` before we attempt a rollback or a proxy shutdown.

```java
proxy.applyConfiguration(newConfig)
     .whenComplete((result, ex) -> {
         if (ex instanceof ConcurrentApplyException) {
             // Another apply is in flight; not a real failure — trigger can retry later.
             return;
         }
         if (result != null && result.hasErrors()) {
             proxy.applyConfiguration(oldConfig)
                  .whenComplete((rollbackResult, rollbackEx) -> {
                      if (rollbackEx != null || rollbackResult.hasErrors()) {
                          // rollback itself failed — last-resort policy
                          proxy.shutdown();
                      }
                  });
         }
     });
```

Have updated the examples and also added a small section under `Concurrency control` to explain how to deal with `ConcurrentApplyException` for any trigger developers

## Proposal

### Core API: `KafkaProxy.applyConfiguration()`

The central operation is:

```java
class KafkaProxy {
    // ... add the following method

    /**
     * Apply the given configuration to this running proxy, restarting only the
     * virtual clusters whose effective configuration differs from the current
     * running state. Unaffected clusters continue serving traffic throughout
     * the apply.
     *
     * <h2>Validation contract</h2>
     * <p>Static validation (schema conformance, required fields, field-value

>      * ranges, internal consistency) is the embedder's responsibility and is

Reviewer 3: ```suggestion
     * ranges, internal consistency) is the caller's responsibility and is
```

     * exceptional completion of the returned future.
     *
     * @param newConfig the desired end-state configuration; must be non-null
     *                  and statically valid
     * @return a future that completes with a {@link ReloadResult} listing which
     *         clusters ended up unchanged, added, restarted, removed, or failed
     * @throws NullPointerException  if {@code newConfig} is {@code null}
     * @throws IllegalStateException if the proxy is not in the running state
     */
    public CompletableFuture<ReloadResult> applyConfiguration(Configuration newConfig);
}
```

The caller provides a complete `Configuration` object. The proxy compares it against the currently running configuration, determines what changed, and applies the changes. The method returns a `CompletableFuture<ReloadResult>` that completes with a structured result on success or exceptionally on failure.

`ReloadResult` reports which clusters ended up in each terminal outcome of the apply:

```java
public record ReloadResult(
    Set<String> clustersUnchanged,
    Set<String> clustersAdded,
    Set<String> clustersRestarted,

>     Set<String> clustersRemoved,

Reviewer 3: You should say that these are all VC names.

Author: You mean to say that we should add relevant javadocs with this information or change the variable naming or maybe both

>     Set<String> failedClusters

Reviewer 3: I suspect eventually we will come to care about why those clusters failed, and so a `Set<String>` of the names will become a bit limiting. Since it's not a public API it would be relatively easy to change it later, but maybe we should make it a `Map<String, Exception>` now, all the same?
